teek/_widgets/treeview.py

```diff
@@ -0,0 +1,469 @@
+        return len(self._data)
+
+    def insert(self, index, row):
+        """
+        Insert a row or values converted to a row at the given index.
+        """
+        # Convert to row
+        if not isinstance(row, TreeviewRow):
+            row = TreeviewRow(values=row)
+
+        self._treeview._call(None, self._treeview, 'insert', '', index, '-id',
+                             row)
+        row._assign(self._treeview)
+        self._data.insert(index, row)
+
+    def move(self, from_pos, to_pos):
+        """
+        Move row at index to other position
+        """
+        row = self._data[from_pos]
+        del self[from_pos]
+        self.insert(to_pos, row)
+
+
+class Treeview(ChildMixin, Widget):
+    """
+    This is the treeview widget.
+
+    :class:`TreeviewRow` and :class:`TreeviewColumn` objects can be added to
+    this widget by calling ``insert`` or ``append`` of the ``rows`` or
+    ``column`` property. In general, these properties behave like a
+    :class:`list` of :class:`TreeviewRow` or :class:`TreeviewColumn` objects.
+
+    If you want to move a row, use the ``move`` method of the ``rows`` property
+    like this:
+    ::
+
+        treeview.rows.move(2, 0) # move row at index 2 to index 0
+
+    Manual page: :man:`ttk_treeview(3tk)`
+    """
+
+    _widget_name = 'ttk::treeview'
+    tk_class_name = 'Treeview'
+
+    def __init__(self, parent, *, rows=[], columns=[], **kwargs):
+        super().__init__(parent, **kwargs)
+
+        self.rows = TreeviewRowList(self, rows)
+        self.columns = TreeviewColumnList(self, columns)
+
+    def _init_config(self):
+        super()._init_config()
+
+        self.config._types.update({
+            'selectmode': str,
+            'height': int,
+            'show': [str]
+        })
+
+    def _repr_parts(self):
+        return ["contains %d columns & %d rows" % (len(self.columns),
+                                                   len(self.rows))]
+
+    def _xview_or_yview(self, xview_or_yview, *args):
+        if not args:
+            return self._call((float, float), self, xview_or_yview)
+
+        self._call(None, self, xview_or_yview, *args)
+        return None
+
+    xview = functools.partialmethod(_xview_or_yview, 'xview')
+    yview = functools.partialmethod(_xview_or_yview, 'yview')
+
+    def sort(self, column_pos, reverse=False):
+        """
+        Sort all rows according to value of column with given ``column_pos``.
+        The direction of the sorting can be set with ``reverse``. If this value
+        is true, the rows are sorted in ascending order.
+        """
+        if column_pos < 1:
+            raise KeyError('column position must be greater or equal 1')
+
+        sorted_rows = self.rows._data.copy()
+        sorted_rows.sort(key=lambda r: r.config['values'][column_pos - 1],
+                         reverse=reverse)
+
+        for old, row in enumerate(self.rows):
+            self.rows.move(old, sorted_rows.index(row))
```

Comment:
Sort is a good idea, and I have seen tkinter users who have wanted it in tkinter.

This code feels wrong to me for a couple reasons:

- You are modifying `self.rows` in a loop that loops over `self.rows`. That is generally a bad idea. For example, what does this code print?

    ```
    stuff = ['a', 'b', 'c', 'd', 'e']
    for thing in stuff:
        stuff.remove(thing)
    print(stuff)
    ```

    The problem is that Python uses an index variable for looping internally, so it does `stuff.remove(stuff[0])`, `stuff.remove(stuff[1])` etc until `stuff[the index]` gives `IndexError`. The list shrinks as items are removed from it, which screws up the indexes; first this removes `a` (element 0), and then because the list is `['b', 'c', 'd', 'e']` it removes `'c'` (element 1), and then the list is `['b', 'd', 'e']` and it removes `'e'` (element 2). A simple solution is to loop over `stuff.copy()`.

- There is a similar issue with `self.rows.move`. Once an element has been moved, its index may still change as other items are moved. For example, if the rows are `[A, B, C]` after moving row `B` to index 1, and row `A` is moved to index 2, then `B` will be at index 0.

Here is a simpler example code that does not sort a list correctly:

```python3
import random

stuff = [7, 3, 2, 5, 6, 1, 8, 4]

def move(old, new):
    value = stuff[old]
    del stuff[old]
    stuff.insert(new, value)

sorted_stuff = sorted(stuff)
for old, value in enumerate(stuff):
    move(old, sorted_stuff.index(value))

print(stuff)
```

I would try something like this

```python3
for row in reversed(sorted_rows):
    self.rows.move(self.rows.index(row), 0)
```

This moves each item to the beginning, moving all the previously moved items after it. This is not ideal either, and it will create a *lot* of moves. There is probably a better way to do this, but at least the `move` method should be implemented with Tk's `move` command (search for `pathname move` in ttk_treeview manual page).

I think it's fine if you delete `sort` from this pull request and let me or someone else add it :)

Reply:
This alternative seems to also work, and it would create less unnecessary moves especially when the treeview is already almost sorted (creates 0 moves if the treeview is already sorted):

```python3
for index, row in enumerate(sorted_rows):
    if self.rows[index] != row:
        self.rows.move(self.rows.index(row, index), index)
```